<html xmlns:concordion="http://www.concordion.org/2007/concordion" xmlns:cc="http://fewi.custom.commands.FewiExtensions">
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<style>
	table {border-collapse: collapse; padding-bottom:5px; margin-bottom:5px;}
	table, th, td {border: 1px solid #CCCCCC;}
	td, th {padding:2px;}
	.man {background-color:#FFFFCC;}
	ul {padding: 0px; padding-bottom:3px; padding-top:0px; margin: 0px;}
	ol {padding: 0px; padding-bottom:3px; padding-top:0px; margin: 0px;}
	li {padding-left:2px; padding-top:0px; padding-bottom:0px; margin-left:25px;}
	p {margin:2px 0px;}
</style>
<!-- declare strict doctype to look for disallowed html errors
 **  **  **  **  ** REMOVE THE LINE before committing the file **  **  **  **  **  **  **
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
 **  **  **  **  ** REMOVE THE LINE before committing the file **  **  **  **  **  **  **
-->


    <title>Marker Summary from Query</title>
</head>


<body>

<div id="TOC">
	<p>Contents: <a href="#maintitle">Title</a> | <a href="#acceptance">Acceptance Criteria</a> | <a href="#definitions">Definitions</a> | <a href="#logic">Logic Summary</a> | <a href="#manual">Manual Tests</a> | <a href="#auto">Automated Tests</a></p>
</div>

<div id="maintitle">
    <h1>Marker Summary from Query</h1>
    <ul>
        <li>User story 163, E9: MQ&amp;S: Search Functionality (MGI 5.18)</li>
        <li>User story 167, E9: MQ&amp;S: Sorting (5.18)</li>
        <li>User story 168, E9: MQ&amp;S: You searched for (5.18)</li>
        <li>User story 164, E9: MQ&amp;S: Text/Excel downloads (5.18)</li>
        <li>User story 199, E9: Boolean enable GO and InterPro fields</li>
        <li>User story 203, E9: Add sort by Feature type to marker summary</li>
        <li>User story 215,	Bug: Breast Cancer search returns Riken genes</li>
    </ul>
</div>


<div id="acceptance">
    <h2>Acceptance Criteria</h2>
	<ul>
		<li>Small differences due to using a different text search engines are OK. Determining if a difference if "small" is mainly a judgement call by the SME, I think.</li>
		<li>nomen search has no operator - as in allele and GXD forms</li>
		<li>verify GO and InterPro work ok</li>
		<li>add pheno system selection box</li>
		<li>other fields should give exactly the same results, e.g., a query by coordinates, or GO term	</li>
		<li>sto167: user can sort by symbol or by location by clicking in the column header</li>
		<li>sto167: symbol sort is smart alpha</li>
		<li>sto167: location sort is (smart) chromosome + startposition + symbol</li>
		<li>sto167: default sort is by symbol</li>
		<li>sto167: special case for query by nomen that promotes mouse matches over other species - need to define exactly when it applies</li>
		<li>sto168: query summary display a "you searched for" section above the table</li>
		<li>sto168: it contains the users search parameters, nicely formatted</li>
		<li>sto168: SME word requirements satisfied</li>		
		<li>sto164: user can download results at tab-delimited text</li>
		<li>sto164: user can download results as an Excel spreadsheet</li>
		<li>sto164: download columns include the same data as the web display PLUS the MGI id</li>
		<li>sto164: download columns are separated (where they are combined in the web display)</li>
		<li>sto199: both the GO and InterPro fields support boolean operators (AND, OR, NOT)</li>
		<li>sto203: Feature Type sort is smart alpha.</li>
		<li>sto215: TEXT-QTL mapping experiment notes are not searched unless the markers are of type QTL for the Pheno/Disease field</li>
	</ul>
</div>


<div id="definitions">
	<h2>Definitions</h2>
	<dl>
		<dt>Term</dt>
		<dd>definition line 1</dd>
		<dd>definition line 2</dd>
	</dl>
</div>


<div id="logic">
	<h2>Logic Summary</h2>

	<h3>Querying in Text Fields</h3>
	<p>Description of logic used in each of the text field inputs when a query is submitted:</p>

	<h4>Nomenclature</h4>
	<ul>
		<li>Performs a tokenized query against Solr index.</li>
		<li>Initial and trailing wildcard symbols (asterisk) are supported; "internal" asterisks are treated literally.</li>
		<li>Tokens are combined with Boolean AND. </li>
		<li>Does not include search against allele synonyms.</li>
		<li>Attempts to recreate, as much as possible, previous code's default sorting if Nomen field is submitted (except now using Solr field boosts):<br/>
			<b>MRK_label table</b>, April 2014, additional Python code sorted "exact" matches ahead of matches contained within text strings.
		<table>
					<tr>
						<th>labeltype</th>
						<th>labeltypename</th>
						<th>organism name</th>
						<th>priority</th>
						<th>new order 4/23/2014</th>
					</tr>
					<tr>
						<td>MS</td>
						<td>current symbol</td>
						<td>mouse</td>
						<td>1</td>
						<td>1</td>
					</tr>
					<tr>
						<td>MY</td>
						<td>synonym</td>
						<td>mouse</td>
						<td>7</td>
						<td>2</td>
					</tr>
					<tr>
						<td>MS</td>
						<td>old symbol</td>
						<td>mouse</td>
						<td>5</td>
						<td>3</td>
					</tr>
					<tr>
						<td>MY</td>
						<td>human synonym</td>
						<td>human</td>
						<td>8</td>
						<td>4</td>
					</tr>
					<tr>
						<td>MN</td>
						<td>current name</td>
						<td>mouse</td>
						<td>2</td>
						<td>5</td>
					</tr>
					<tr>
						<td>AS</td>
						<td>allele symbol</td>
						<td>mouse</td>
						<td>3</td>
						<td>6</td>
					</tr>
					<tr>
						<td>AN</td>
						<td>allele name</td>
						<td>mouse</td>
						<td>4</td>
						<td>7</td>
					</tr>
					<tr>
						<td>MN</td>
						<td>old name</td>
						<td>mouse</td>
						<td>6</td>
						<td>8</td>
					</tr>
					<tr>
						<td>MY</td>
						<td>rat synonym</td>
						<td>rat</td>
						<td>9</td>
						<td>9</td>
					</tr>
					<tr>
						<td>MY</td>
						<td>related synonym</td>
						<td></td>
						<td>10</td>
						<td>10</td>
					</tr>
					<tr>
						<td>OS</td>
						<td>human symbol</td>
						<td>human</td>
						<td>11</td>
						<td>11</td>
					</tr>
					<tr>
						<td>ON</td>
						<td>human name</td>
						<td>human</td>
						<td>12</td>
						<td>12</td>
					</tr>
					<tr>
						<td>OS</td>
						<td>rat symbol</td>
						<td>rat</td>
						<td>13</td>
						<td>13</td>
					</tr>
					<tr>
						<td>OS</td>
						<td>rhesus macaque symbol</td>
						<td>rhesus macaque</td>
						<td>14</td>
						<td>14</td>
					</tr>
					<tr>
						<td>OS</td>
						<td>cattle symbol</td>
						<td>cattle</td>
						<td>14</td>
						<td>14</td>
					</tr>
					<tr>
						<td>OS</td>
						<td>dog symbol</td>
						<td>dog, domestic</td>
						<td>14</td>
						<td>14</td>
					</tr>
					<tr>
						<td>OS</td>
						<td>zebrafish symbol</td>
						<td>zebrafish</td>
						<td>14</td>
						<td>14</td>
					</tr>
					<tr>
						<td>OS</td>
						<td>chicken symbol</td>
						<td>chicken</td>
						<td>14</td>
						<td>14</td>
					</tr>
					<tr>
						<td>OS</td>
						<td>chimpanzee symbol</td>
						<td>chimpanzee</td>
						<td>14</td>
						<td>14</td>
					</tr>
				</table>
			</li>
	</ul>

	<h4>GO and Protein Domain (InterPro)</h4>
	<ul>
		<li>Performs a tokenized query against Solr index.</li>
		<li>Performs stemming.</li>
		<li>Supports initial and trailing wildcards.</li>
		<li>Supports quotation marks for phrase matching.</li>
		<li>Supports parentheses for grouping.</li>
		<li>Supports Booleans AND, OR, NOT, AND NOT. Default is OR.</li>
		<li>The 3 GO DAG root node terms are excluded.</li>
	</ul>

	<h4>Pheno/Disease</h4>
	<ul>
		<li>Performs a tokenized query against Solr index.</li>
		<li>Initial and trailing wildcard symbols (asterisk) are supported.</li>
		<li>Supports quotation marks for phrase matching.</li>
		<li>Supports parentheses for grouping.</li>
		<li>Supports Booleans AND, OR, NOT, AND NOT. Default is OR.</li>
		<li>Returns only genes having annotated alleles in One Marker Genotypes (OMG) or "Simple" genotypes: 2-marker Conditional genotypes and 2-marker genotypes involving a Transgenic Reporter.
		<ul>
			<li>EXCLUDES markers for annotated alleles present only in Complex genotypes (unlike the Allele Query): genotypes having 3 or more markers; genotypes having 2 markers that are not Conditional and do not contain a Transgenic Reporter.</li>
			<li>EXCLUDES markers of slash alleles (unlike Allele Query) - examples:
			<ul>
				<li>MP:0004800 returns Lta/Tnf&lt;tm1Eug&gt; twice and also Lta/Tnf&lt;tm1Jods&gt; twice on the Allele Query, but neither Lta nor Tnf on the Marker Query.</li>
				<li>MP:0004984 returns Bglap/Bglap2&lt;tm1Kry&gt; twice on the Allele Query, but neither Bglap nor Bglap2 on the Marker Query.</li>
			</ul>
			</li>
		</ul>
		</li>
		<li>Excludes Recombinase markers in Conditional genotypes.</li>
		<li>Excludes Transgenic Reporter markers.</li> 
		<li>Searches for matches in Mapping Experiment Notes when the Mapping Experiment type is TEXT-QTL only for markers of type QTL.</li>
	</ul>
</div>




<div id="tests">
    <h2>Tests</h2>
    <p>Skip to <a href="#manual">Manual Tests</a> |  <a href="#auto">Automated Tests</a></p>
    <p><a href="#nomenclature">Nomenclature</a> |  <a href="#mapPosition">Map position</a> | <a href="#goClassification">Gene Ontology classifications</a> | <a href="#proteinDomains">Protein domains</a> | <a href="#mousePheno">Mouse phenotypes &amp; mouse models of human disease</a> | <a href="#sorting">Sorting tests</a> | <a href="#youSearchedFor">You searched for</a> | <a href="#downloadableFiles">Downloadable Files</a> </p>
	<div id="manual">
<!-- To insert manual test results, use the following:
<p concordion:manual="true">testing assertTrue - displays as green</p>
<p concordion:manual="false">testing assertFalse - displays as red</p>
 -->
		<h3>Manual tests</h3>
		
		<h4 id="nomenclature">Nomenclature</h4>
		<ol>
    		<li>Current, old, and interim mouse marker symbols, names, and synonyms  and human ortholog symbols and names are supported.
 			<p concordion:manual="true">passed.  -klf 3/31/2014</p>
    		<ol>
                <li>Navigate to the marker query form. Search for the following and verify the indicated results are returned.</li>
                <li>
                <table>
								<tr>
									<th>Input<br /> <span style="color: #FFFFFF;">ForceCellToBeWider</span></th>
									<th>Include in return<br />(may be others)
									</th>
									<th>Exclude from return</th>
									<th>Notes to tester</th>
								</tr>
								<tr>
									<td>Pax6</td>
									<td>
										<ul>
											<li>Pax6</li>
											<li>Paupar</li>
											<li>Del(2)Pax6<sup>11Neu</sup></li>
											<li>Tg(CAG-EGFP,-Pax6*5a,-lacZ)1Stoy</li>
										</ul>
									</td>
									<td>
										<ul>
											<li>Pax6os1</li>
											<li>Elp4</li>
										</ul>
									</td>
									<td>Pax6os1 and Elp4 are no longer returned because we no
										longer do default "contains" searching within tokens. These
										markers are returned via Current Symbol, Current Name and
										Synonym.</td>
								</tr>
								<tr>
									<td>Defb1</td>
									<td>Defb1</td>
									<td></td>
									<td>This is an interim marker, searching by Symbol.</td>
								</tr>
								<tr>
									<td>pentamer</td>
									<td>Ppnr</td>
									<td></td>
									<td>This is an interim marker, searching by a token in the
										Name.</td>
								</tr>
								<tr>
									<td>oral cancer</td>
									<td>
										<ul>
											<li>Cdk2ap1</li>
											<li>Oraov1</li>
										</ul>
									</td>
									<td></td>
									<td>Cdk2ap1 is returned via Old Name</td>
								</tr>
								<tr>
									<td>asip</td>
									<td>
										<ul>
											<li>a</li>
											<li>Pard3</li>
										</ul>
									</td>
									<td></td>
									<td>Ortholog has different symbol.</td>
								</tr>
								<tr>
									<td>&quot;CD3-TCR complex&quot;</td>
									<td>
										<ul>
											<li>Cd3d</li>
											<li>Cd3e</li>
											<li>Cd3g</li>
										</ul>
									</td>
									<td></td>
									<td>Query using quotation marks. Ortholog has different
										name.</td>
								</tr>
								<tr>
									<td>"RT1-Db1"</td>
									<td>H2-Eb1</td>
									<td></td>
									<td>Rat ortholog.</td>
								</tr>
								<tr>
									<td>Acta1a</td>
									<td>Actg2</td>
									<td></td>
									<td>zebrafish symbol</td>
								</tr>
								<tr>
									<td>LOC776816 </td>
									<td>Actg1</td>
									<td></td>
									<td>chicken</td>
								</tr>
								<tr>
									<td>RPS4Y1 </td>
									<td>Rps4x</td>
									<td></td>
									<td>cattle</td>
								</tr>
								<tr>
									<td>ELA2B </td>
									<td>Cela2a</td>
									<td></td>
									<td>rhesus macaque</td>
								</tr>
							</table>
						</li>
    		</ol>
    		</li>
    		
    		<li>Current allele symbols or names are supported.  Allele synonyms are NOT supported.
 			<p concordion:manual="true">passed. -klf 3/31/2014</p>
    		<ol>
                <li>Navigate to the marker query form. Search for the following and verify the indicated results are returned.</li>
                <li>
                <table>
								<tr>
									<th>Input<br /> <span style="color: #FFFFFF;">ForceCellToBeWider</span></th>
									<th>Include in return<br />(may be others)
									</th>
									<th>Exclude from return</th>
									<th>Notes to tester</th>
								</tr>
								<tr>
									<td>Gena265</td>
									<td></td>
									<td>Pax6</td>
									<td>Gena265 is a synonym for the allele Pax6<sup>Leca3</sup>.
									</td>
								</tr>
								<tr>
									<td>subcapsularis</td>
									<td></td>
									<td>Mip</td>
									<td>Allele Synonym : cataracta hereditaria subcapsularis
										for Mip<sup>cat</sup>.
									</td>
								</tr>
								<tr>
									<td>leprdb</td>
									<td>Lepr</td>
									<td></td>
									<td>Lepr<sup>db</sup> gets tokenized to leprdb</td>
								</tr>
								<tr>
									<td>corneal adhesion</td>
									<td>Pax6</td>
									<td></td>
									<td>Allele name for Pax6<sup>Leca3</sup></td>
								</tr>
								<tr>
									<td>Leca3</td>
									<td>Pax6</td>
									<td></td>
									<td>Allele name for Pax6<sup>Leca3</sup></td>
								</tr>
							</table>
						</li>
    		</ol>
    		</li>

    		<li>Leading and trailing asterisks are treated as wildcards.  "Internal" asterisks are treated as literal characters.
 			<p concordion:manual="true">passed. -klf 3/31/2014</p>
    		<ol>
                <li>Navigate to the marker query form. Search for the following and verify the indicated results are returned.</li>
                <li>
                <table>
								<tr>
									<th>Input<br /> <span style="color: #FFFFFF;">ForceCellToBeWider</span></th>
									<th>Include in return<br/>(may be others)</th>
									<th>Exclude from return</th>
									<th>Notes to tester</th>
								</tr>
								<tr>
									<td>Pax6*</td>
									<td>
										<ul>
											<li>Pax6</li>
											<li>Pax6os1</li>
											<li>Del(2)Pax6<sup>11Neu</sup></li>
											<li>Elp4</li>
											<li>Paupar</li>
											<li>Tg(Pax6-cre,GFP)1Pgr</li>
										</ul>
									</td>
									<td></td>
									<td></td>
								</tr>
								<tr>
									<td>*os*</td>
									<td>
									<ul>
										<li>Abhd11os</li>
										<li>cosa</li>
										<li>Mid1</li>
										<li>Os</li>
										<li>Os9</li>
										<li>Osbp</li>
										<li>Tg(Etn2*,Ptf1aos,Ptf1a)#Kymm</li>
										<li>Tg(Fos-EGFP)KR563Gsat</li>
									</ul>
									</td>
									<td></td>
									<td></td>
								</tr>
							</table>
						</li>
    		</ol>
    		</li>

    		<li>Non-alphanumeric characters other than asterisks are used to define token boundaries.
 			<p concordion:manual="true">passed. -klf 3/31/2014</p>
    		<ol>
                <li>Navigate to the marker query form. Search for the following and verify the indicated results are returned.</li>
                <li>
                	<table>
						<tr>
							<th>Input<br /> <span style="color: #FFFFFF;">ForceCellToBeWider</span></th>
							<th>Include in return<br />(may be others)
							</th>
							<th>Exclude from return</th>
							<th>Notes to tester</th>
						</tr>
						<tr>
							<td>Dach1-Pax6</td>
							<td>Tg(Dach1-Pax6)#Fjl</td>
							<td></td>
							<td></td>
						</tr>
						<tr>
							<td>-Pax6+Dach1</td>
							<td>Tg(Dach1-Pax6)#Fjl</td>
							<td></td>
							<td>Not a string in a real name or symbol; verifying that the tokens are ANDed and characters don't matter.</td>
						</tr>
						<tr>
							<td>Pax6*5a</td>
							<td>Tg(CAG-EGFP,-Pax6*5a,-lacZ)1Stoy</td>
							<td></td>
							<td></td>
						</tr>
						<tr>
							<td>-Pax6,-lacZ</td>
							<td>
								<ul>
									<li>Tg(CAG-EGFP,-Pax6*5a,-lacZ)1Stoy</li>
									<li>Tg(CAG-EGFP,-Pax6,-lacZ)1Stoy</li>
								</ul>
							</td>
							<td></td>
							<td></td>
						</tr>
						<tr>
							<td>BV100Gsat</td>
							<td>Tg(Pax6-EGFP)BV100Gsat</td>
							<td></td>
							<td></td>
						</tr>
						<tr>
							<td>Gsat</td>
							<td></td>
							<td>Tg(Pax6-EGFP)BV100Gsat</td>
							<td>Verifying that tokenization is not happening on character-numeric string changes.</td>
						</tr>
					</table>
				</li>
    		</ol>
    		</li>
		</ol>

   		<h4 id="mapPosition">Map position</h4>
		<ol>
    		<li>Submitting an invalid marker range (one or both markers are missing coordinates; markers are on different chromosomes) returns error messages.
 			<p concordion:manual="true">passed. -klf 3/31/2014</p>
 			<ol>
 				<li>Input <b>Aff3</b> and <b>Tbx10</b>.</li>
 				<li>Verify result returns <b>Marker Aff3(1) not on same chromosome as Tbx10(19)</b>.</li>
 				<li>Input <b>Tbx10</b> and <b>D19Mit127</b>.</li>
 				<li>Verify result returns <b>end marker D19Mit127 has coordinates</b></li>
 				<li>Input <b>D19Mit127</b> and <b>Tbx10</b>.</li>
 				<li>Verify result returns <b>start marker D19Mit127 has coordinates</b></li>
 			</ol>
 			</li>

		</ol>

   		<h4 id="goClassification">Gene Ontology classifications</h4>
		<ol>
    		<li>Submitting a GO ID returns markers associated with the term and its descendants, if any.
    		<ol>
                <li>Enter <b concordion:set="#go">GO:0060710</b> (chorio-allantoic fusion) and submit the query. (has no descendants)</li>
                <li>Verify the genes returned include:
                <table cc:verifySupersetOf="#symbol : getSymbolsByGo(#go)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>Cyr61</td></tr>
                	<tr><td>Dnajb6</td></tr>
                	<tr><td>Erf</td></tr>
                	<tr><td>Itga4</td></tr>
                	<tr><td>Lef1</td></tr>
                	<tr><td>Vcam1</td></tr>
                	<tr><td>Wnt7b</td></tr>
                </table>
                </li>
                <li>Enter <b concordion:set="#go">GO:0070486</b> (leukocyte aggregation) and submit the query. (has annotated descendants; only Sema4d is annotated to the parent term)</li>
                <li>Verify the genes returned include:
                <table cc:verifySupersetOf="#symbol : getSymbolsByGo(#go)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>Adam8</td></tr>
                	<tr><td>Bmp7</td></tr>
                	<tr><td>Cd44</td></tr>
                	<tr><td>Gstp1</td></tr>
                	<tr><td>Has2</td></tr>
                	<tr><td>Il1b</td></tr>
                	<tr><td>S100a8</td></tr>
                	<tr><td>S100a9</td></tr>
                	<tr><td>Sema4d</td></tr>
                	<tr><td>Spn</td></tr>
                	<tr><td>Stk10</td></tr>
                </table>
                </li>
    		</ol>
    		</li>

    		<li>Submitting an exact GO term returns markers associated with the term and its descendants, if any.
    		<ol>
                <li>Enter <b concordion:set="#go">leukocyte aggregation</b> (GO:0070486) and submit the query. (has annotated descendants; only Sema4d is annotated to the parent term)</li>
                <li>Verify the genes returned include:
                <table cc:verifySupersetOf="#symbol : getSymbolsByGo(#go)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>Adam8</td></tr>
                	<tr><td>Bmp7</td></tr>
                	<tr><td>Cd44</td></tr>
                	<tr><td>Gstp1</td></tr>
                	<tr><td>Has2</td></tr>
                	<tr><td>Il1b</td></tr>
                	<tr><td>S100a8</td></tr>
                	<tr><td>S100a9</td></tr>
                	<tr><td>Sema4d</td></tr>
                	<tr><td>Spn</td></tr>
                	<tr><td>Stk10</td></tr>
                </table>
                </li>
    		</ol>
    		</li>

    		<li>Submitting an exact GO synonym returns markers associated with the term and its descendants, if any.
    		<ol>
                <li>Enter <b concordion:set="#go">leucocyte aggregation</b> (synonym for GO:0070486) and submit the query. (has annotated descendants; only Sema4d is annotated to the parent term)</li>
                <li>Verify the genes returned include:
                 <table cc:verifySupersetOf="#symbol : getSymbolsByGo(#go)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>Adam8</td></tr>
                	<tr><td>Bmp7</td></tr>
                	<tr><td>Cd44</td></tr>
                	<tr><td>Gstp1</td></tr>
                	<tr><td>Has2</td></tr>
                	<tr><td>Il1b</td></tr>
                	<tr><td>S100a8</td></tr>
                	<tr><td>S100a9</td></tr>
                	<tr><td>Sema4d</td></tr>
                	<tr><td>Spn</td></tr>
                	<tr><td>Stk10</td></tr>
                </table>
                </li>
    		</ol>
    		</li>

    		<li>Submitting text which appears in more than one GO term returns markers associated with each term and its descendants, if any.
    		<ol>
                <li>Enter <b concordion:set="#go">reticulum</b> in the GO input text field and submit the search.<br/>(Appears in 3 Molecular Function, 96 Biological Process, and 64 Cellular Component terms.) </li>
                <li>Verify the genes returned include the following:
	                <table
								cc:verifySupersetOf="#symbol : getSymbolsByGo(#go)">
								<tr>
									<th concordion:assertEquals="#symbol">Gene</th>
									<th>Annotated Term</th>
									<th>GO sub-DAG</th>
								</tr>
								<tr>
									<td>Srp14</td>
									<td>endoplasmic reticulum signal peptide binding<br />
										establishment of protein localization to endoplasmic reticulum<br />
										signal recognition particle, endoplasmic reticulum targeting
									</td>
									<td>Molecular Function<br /> Biological Process<br />
										Cellular Component
									</td>
								</tr>
								<tr>
									<td>Srp54a</td>
									<td>endoplasmic reticulum signal peptide binding<br />
										establishment of protein localization to endoplasmic reticulum<br />
										signal recognition particle, endoplasmic reticulum targeting
									</td>
									<td>Molecular Function<br /> Biological Process<br />
										Cellular Component
									</td>
								</tr>
								<tr>
									<td>Atp2b4</td>
									<td>sarco(endo)plasmic reticulum Ca2+-ATPase</td>
									<td>Molecular Function</td>
								</tr>
								<tr>
									<td>Cemip</td>
									<td>endoplasmic reticulum retention sequence binding<br/>
										endoplasmic reticulum</td>
									<td>Molecular Function<br />
										Cellular Component</td>
								</tr>
								<tr>
									<td>Rab10</td>
									<td>endoplasmic reticulum tubular network organization<br />
										endoplasmic reticulum
									</td>
									<td>Biological Process<br /> Cellular Component
									</td>
								</tr>
								<tr>
									<td>Reep1</td>
									<td>endoplasmic reticulum tubular network organization<br />
										endoplasmic reticulum
									</td>
									<td>Biological Process<br /> Cellular Component
									</td>
								</tr>
								<tr>
									<td>Casq1</td>
									<td>endoplasmic reticulum organization<br /> sarcoplasmic
										reticulum
									</td>
									<td>Biological Process<br /> Cellular Component
									</td>
								</tr>
								<tr>
									<td>H13</td>
									<td>integral component of cytoplasmic side of endoplasmic
										reticulum membrane</td>
									<td>Cellular Component</td>
								</tr>
								<tr>
									<td>Pkd2</td>
									<td>integral component of cytoplasmic side of endoplasmic
										reticulum membrane</td>
									<td>Cellular Component</td>
								</tr>
								<tr>
									<td>Jph1</td>
									<td>endoplasmic reticulum</td>
									<td>Cellular Component</td>
								</tr>
							</table>
						</li>
    		</ol>
    		</li>
    		
    		<li>Submitting text and checking a box for particular DAG(s) within the GO returns markers associated with matching terms in only the checked DAG(s). RECHECKED FOR STO199.  NEED TO UPDATE WITH DATA FROM TABLE ABOVE.
 			<p concordion:manual="true">passed. -klf 4/16/2014</p>
    		<ol>
                <li>Enter <b>reticulum</b> in the GO input text field.<br/>(Appears in 3 Molecular Function, 96 Biological Process, and 64 Cellular Component terms.) </li>
				<li>Check only the box for <b>Molecular Function</b> and submit the search.</li>
                <li>Verify the genes returned include or exclude the following, as indicated:
	            <table>
					<tr>
						<th>Gene</th>
						<th>Annotated Term</th>
						<th>Included / Excluded</th>
						<th>GO sub-DAG</th>
					</tr>
					<tr>
						<td>Srp14</td>
						<td>endoplasmic reticulum signal peptide binding</td>
						<td>included</td>
						<td>Molecular Function</td>
					</tr>
					<tr>
						<td>Srp54a</td>
						<td>endoplasmic reticulum signal peptide binding</td>
						<td>included</td>
						<td>Molecular Function</td>
					</tr>
					<tr>
						<td>Rab10</td>
						<td>endoplasmic reticulum tubular network organization</td>
						<td>excluded</td>
						<td>Biological Process</td>
					</tr>
					<tr>
						<td>Reep1</td>
						<td>endoplasmic reticulum tubular network organization</td>
						<td>excluded</td>
						<td>Biological Process</td>
					</tr>
					<tr>
						<td>H13</td>
						<td>integral component of cytoplasmic side of endoplasmic
							reticulum membrane</td>
						<td>excluded</td>
						<td>Cellular Component</td>
					</tr>
					<tr>
						<td>Pkd2</td>
						<td>integral component of cytoplasmic side of endoplasmic
							reticulum membrane</td>
						<td>excluded</td>
						<td>Cellular Component</td>
					</tr>
				</table>
				</li>
                <li>Enter <b>reticulum</b> in the GO input text field.<br/>(Appears in 3 Molecular Function, 96 Biological Process, and 64 Cellular Component terms.) </li>
				<li>Check only the box for <b>Biological Process</b> and submit the search.</li>
                <li>Verify the genes returned include or exclude the following, as indicated:
                <table>
					<tr>
						<th>Gene</th>
						<th>Annotated Term</th>
						<th>Included / Excluded</th>
						<th>GO sub-DAG</th>
					</tr>
					<tr>
						<td>Srp14</td>
						<td>endoplasmic reticulum signal peptide binding</td>
						<td>included</td>
						<td>Molecular Function</td>
					</tr>
					<tr>
						<td>Srp54a</td>
						<td>endoplasmic reticulum signal peptide binding</td>
						<td>included</td>
						<td>Molecular Function</td>
					</tr>
					<tr>
						<td>Rab10</td>
						<td>endoplasmic reticulum tubular network organization</td>
						<td>included</td>
						<td>Biological Process</td>
					</tr>
					<tr>
						<td>Reep1</td>
						<td>endoplasmic reticulum tubular network organization</td>
						<td>included</td>
						<td>Biological Process</td>
					</tr>
					<tr>
						<td>H13</td>
						<td>integral component of cytoplasmic side of endoplasmic
							reticulum membrane</td>
						<td>excluded</td>
						<td>Cellular Component</td>
					</tr>
					<tr>
						<td>Pkd2</td>
						<td>integral component of cytoplasmic side of endoplasmic
							reticulum membrane</td>
						<td>excluded</td>
						<td>Cellular Component</td>
					</tr>
				</table>
				</li>
                <li>Enter <b>reticulum</b> in the GO input text field.<br/>(Appears in 3 Molecular Function, 96 Biological Process, and 64 Cellular Component terms.) </li>
				<li>Check only the box for <b>Cellular Component</b> and submit the search.</li>
                <li>Verify the genes returned include or exclude the following, as indicated:
                <table>
					<tr>
						<th>Gene</th>
						<th>Annotated Term</th>
						<th>Included / Excluded</th>
						<th>GO sub-DAG</th>
					</tr>
					<tr>
						<td>Srp14</td>
						<td>endoplasmic reticulum signal peptide binding</td>
						<td>included</td>
						<td>Molecular Function</td>
					</tr>
					<tr>
						<td>Srp54a</td>
						<td>endoplasmic reticulum signal peptide binding</td>
						<td>included</td>
						<td>Molecular Function</td>
					</tr>
					<tr>
						<td>Rab10</td>
						<td>endoplasmic reticulum tubular network organization</td>
						<td>included</td>
						<td>Biological Process</td>
					</tr>
					<tr>
						<td>Reep1</td>
						<td>endoplasmic reticulum tubular network organization</td>
						<td>included</td>
						<td>Biological Process</td>
					</tr>
					<tr>
						<td>H13</td>
						<td>integral component of cytoplasmic side of endoplasmic
							reticulum membrane</td>
						<td>included</td>
						<td>Cellular Component</td>
					</tr>
					<tr>
						<td>Pkd2</td>
						<td>integral component of cytoplasmic side of endoplasmic
							reticulum membrane</td>
						<td>included</td>
						<td>Cellular Component</td>
					</tr>
				</table>
				</li>
                <li>Enter <b>reticulum</b> in the GO input text field.<br/>(Appears in 3 Molecular Function, 96 Biological Process, and 64 Cellular Component terms.) </li>
				<li>Check only the boxes for <b>Molecular Function</b> and <b>Cellular Component</b> and submit the search.</li>
                <li>Verify the genes returned include or exclude the following, as indicated:
                <table>
					<tr>
						<th>Gene</th>
						<th>Annotated Term</th>
						<th>Included / Excluded</th>
						<th>GO sub-DAG</th>
					</tr>
					<tr>
						<td>Srp14</td>
						<td>endoplasmic reticulum signal peptide binding</td>
						<td>included</td>
						<td>Molecular Function</td>
					</tr>
					<tr>
						<td>Srp54a</td>
						<td>endoplasmic reticulum signal peptide binding</td>
						<td>included</td>
						<td>Molecular Function</td>
					</tr>
					<tr>
						<td>Rab10</td>
						<td>endoplasmic reticulum tubular network organization</td>
						<td>excluded</td>
						<td>Biological Process</td>
					</tr>
					<tr>
						<td>Reep1</td>
						<td>endoplasmic reticulum tubular network organization</td>
						<td>included</td>
						<td>Biological Process</td>
					</tr>
					<tr>
						<td>H13</td>
						<td>integral component of cytoplasmic side of endoplasmic
							reticulum membrane</td>
						<td>included</td>
						<td>Cellular Component</td>
					</tr>
					<tr>
						<td>Pkd2</td>
						<td>integral component of cytoplasmic side of endoplasmic
							reticulum membrane</td>
						<td>included</td>
						<td>Cellular Component</td>
					</tr>
				</table>
				</li>
    		</ol>
    		</li>

    		<li>The GO input field supports Boolean operators. (sto199)
 			<p concordion:manual="true">passed. -klf 2014-04-16</p>
 			<ol>
 				<li>Input <b>GO:0045009 OR GO:0061343</b> and submit the query. <br/>(GO:0045009 - chitosome - 4 markers annotated to a child term "melanosome membrane" GO:0033162; <br/>GO:0061343 - cell adhesion involved in heart morphogenesis - 2 markers annotated to child term "regulation of cell adhesion involved in heart morphogenesis" GO:0061344.)</li>
 				<li>Verify that <b>ONLY the following</b> genes are returned:
 				<table>
 					<tr><th>Symbol</th></tr>
 					<tr><td>Gpr143</td></tr>
 					<tr><td>Oca2</td></tr>
 					<tr><td>Th</td></tr>
 					<tr><td>Tyrp1</td></tr>
 					<tr><td>Notch1</td></tr>
 					<tr><td>Rbpj</td></tr>
 				</table>
 				</li>
 				<li>Input <b>GO:0045009 AND GO:0061343</b> and submit the query. <br/>(GO:0045009 - chitosome - 4 markers annotated to a child term "melanosome membrane" GO:0033162; <br/>GO:0061343 - cell adhesion involved in heart morphogenesis - 2 markers annotated to child term "regulation of cell adhesion involved in heart morphogenesis" GO:0061344.)</li>
 				<li>Verify that <b>NO genes</b> are returned.</li>
 				<li>Input <b>GO:0031670 AND NOT GO:0071295</b> and submit the query. <br/>(GO:0031670 - cellular response to nutrient - 17 markers annotated to this and child terms; <br/>GO:0071295 - cellular response to vitamin - 7 markers annotated to this and child terms;<br/> GO:0071295 is a child of GO:0031670 so this query should exclude 7 markers brought back by just GO:0031670.)</li>
 				<li>Verify that <b>ONLY the following</b> genes are returned.
 				<table>
 					<tr><th>Symbol</th></tr>
 					<tr><td>Cdkn2b</td></tr>
 					<tr><td>Chmp1a</td></tr>
 					<tr><td>Foxa2</td></tr>
 					<tr><td>Ncoa1</td></tr>
 					<tr><td>Nr1h4</td></tr>
 					<tr><td>Pdk2</td></tr>
 					<tr><td>Rrp8</td></tr>
 					<tr><td>Srf</td></tr>
 					<tr><td>Usf1</td></tr>
 					<tr><td>Usf2</td></tr>
 				</table>
 				</li>
 				<li>Verify that the following genes <b>ARE NOT</b> returned.
 				<table>
 					<tr><th>Symbol</th></tr>
 					<tr><td>Cyp27b1</td></tr>
 					<tr><td>Gas6</td></tr>
 					<tr><td>Med1</td></tr>
 					<tr><td>Snai2</td></tr>
 					<tr><td>Snw1</td></tr>
 					<tr><td>Trim24</td></tr>
 					<tr><td>Vdr</td></tr>
 				</table>
 				</li>
 				<li>Input <b>("cellular response to vitamin") AND  NOT (d OR k OR b9 OR k3)</b> and submit the query. <br/>(GO:0031670 - cellular response to nutrient - 17 markers annotated to this and child terms; <br/>GO:0071295 - cellular response to vitamin - 7 markers annotated to this and child terms;<br/> GO:0071295 is a child of GO:0031670 so this query should exclude 7 markers brought back by just GO:0031670.)</li>
 				<li>Verify that the following genes <b>ARE NOT</b> returned.
 				<table>
								<tr>
									<th>Symbol</th>
								</tr>
								<tr>
									<td>Cyp27b1</td>
								</tr>
								<tr>
									<td>Gas6</td>
								</tr>
								<tr>
									<td>Med1</td>
								</tr>
								<tr>
									<td>Snai2</td>
								</tr>
								<tr>
									<td>Snw1</td>
								</tr>
								<tr>
									<td>Trim24</td>
								</tr>
								<tr>
									<td>Vdr</td>
								</tr>
								<tr>
									<td>Bche</td>
								</tr>
								<tr>
									<td>Cbs</td>
								</tr>
								<tr>
									<td>Dcps</td>
								</tr>
								<tr>
									<td>Lrp6</td>
								</tr>
								<tr>
									<td>Ndor1</td>
								</tr>
							</table>
						</li>
 			</ol>
    		</li>
    		
    		<li>The GO input field supports leading and trailing wildcards (asterisk). (consequence of sto199)
 			<p concordion:manual="true">passed. -klf 2014-04-16</p>
 			<ol>
 				<li>Enter <b>postheparin</b> in the GO input field and submit the query.</li>
 				<li>Note the number of genes returned (16 as of 4/16/2014).</li>
 				<li>Enter <b>posthepar*</b> in the GO input field and submit the query.</li>
 				<li>Verify that the same number of genes is returned.</li>
 				<li>Enter <b>*stheparin</b> in the GO input field and submit the query.</li>
 				<li>Verify that the same number of genes is returned.</li>
 				<li>Enter <b>*sthepar*</b> in the GO input field and submit the query.</li>
 				<li>Verify that the same number of genes is returned.</li>
 			</ol>
    		</li>
    		
    		<li>The GO input field supports grouping using parentheses. (consequence of sto199)
 			<p concordion:manual="true">passed. -klf 2014-04-16</p>
 			<ol>
 				<li>See results of "AND NOT" text-searching test, above.</li>
 			</ol>
    		</li>
    		
    		<li>The GO input field supports phrase matching using double quotation marks. (consequence of sto199)
 			<p concordion:manual="true">passed. -klf 2014-04-16</p>
 			<ol>
 				<li>See results of "AND NOT" text-searching test, above.</li>
 			</ol>
 			</li>


    		
		</ol>
   		<h4 id="proteinDomains">Protein domains</h4>
		<ol>
    		<li>Submitting an InterPro ID returns markers associated with the term (we do not represent InterPro as a DAG or hierarchy, even though it is one).
    		<ol>
                <li>Enter <b concordion:set="#interpro">IPR000001</b> (Kringle) and submit the query. (has no descendants)</li>
                <li>Verify <b>15</b> genes are returned.</li>
                <li>Verify the genes returned include:
	            <table cc:verifySupersetOf="#symbol : getSymbolsByInterpro(#interpro)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>F12</td></tr>
                	<tr><td>F2</td></tr>
                	<tr><td>Habp2</td></tr>
                	<tr><td>Hgfac</td></tr>
                	<tr><td>Mst1</td></tr>
                	<tr><td>Plau</td></tr>
                	<tr><td>Ror1</td></tr>
                	<tr><td>Ror2</td></tr>
                </table>
                </li>
                <li>Enter <b concordion:set="#interpro">IPR000562</b> (Fibronectin, type II, collagen-binding) and submit the query. (has no descendants)</li>
                <li>Verify <b>13</b> genes are returned.</li>
                <li>Verify the genes returned include:
	            <table cc:verifySupersetOf="#symbol : getSymbolsByInterpro(#interpro)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>Bsph1</td></tr>
                	<tr><td>F12</td></tr>
                	<tr><td>Hgfac</td></tr>
                	<tr><td>Ly75</td></tr>
                	<tr><td>Mmp2</td></tr>
                	<tr><td>Mmp9</td></tr>
                	<tr><td>Pla2r1</td></tr>
                	<tr><td>Sel1l</td></tr>
                </table>
                </li>
                <li>Enter <b concordion:set="#interpro">IPR013806</b> (Kringle-like fold) and submit the query. (has annotated descendants; is the parent term of the previous two queried terms; has no genes annotated ONLY to the parent term)</li>
                <li>Verify <b>25</b> genes are returned.</li>
                <li>Verify the genes returned include:
	            <table cc:verifySupersetOf="#symbol : getSymbolsByInterpro(#interpro)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>Bsph1</td></tr>
                	<tr><td>F12</td></tr>
                	<tr><td>F2</td></tr>
                	<tr><td>Habp2</td></tr>
                	<tr><td>Hgfac</td></tr>
                	<tr><td>Mmp2</td></tr>
                	<tr><td>Mmp9</td></tr>
                	<tr><td>Mst1</td></tr>
                	<tr><td>Plau</td></tr>
                	<tr><td>Ror1</td></tr>
                	<tr><td>Ror2</td></tr>
                	<tr><td>Sel1l</td></tr>
                </table>
                </li>
    		</ol>
    		</li>

    		<li>Submitting an exact InterPro term returns markers associated with the term.
    		<ol>
                <li>Enter <b concordion:set="#interpro">Kringle-like fold</b> (IPR013806) and submit the query. (has annotated descendants; is the parent term of the previous two queried terms; has no genes annotated ONLY to the parent term)</li>
                <li>Verify <b>25</b> genes are returned.</li>
                <li>Verify the genes returned include:
                <table cc:verifySupersetOf="#symbol : getSymbolsByInterpro(#interpro)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
                	<tr><td>Bsph1</td></tr>
                	<tr><td>F12</td></tr>
                	<tr><td>F2</td></tr>
                	<tr><td>Habp2</td></tr>
                	<tr><td>Hgfac</td></tr>
                	<tr><td>Mmp2</td></tr>
                	<tr><td>Mmp9</td></tr>
                	<tr><td>Mst1</td></tr>
                	<tr><td>Plau</td></tr>
                	<tr><td>Ror1</td></tr>
                	<tr><td>Ror2</td></tr>
                	<tr><td>Sel1l</td></tr>
                </table>
                </li>
    		</ol>
    		</li>

    		<li>Submitting text which appears in more than one InterPro term returns markers associated with any matching term(s).
    		<ol>
                <li>Enter <b concordion:set="#interpro">fibronectin</b> (piece of text which appears in many protein domain names which have annotated genes) and submit the query.</li>
                <li>Verify <b>231</b> genes are returned.</li>
                <li>Verify the genes returned include:
	            <table cc:verifySupersetOf="#symbol : getSymbolsByInterpro(#interpro)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
        			<tr><td>Abi3bp</td></tr>
                	<tr><td>Bsph1</td></tr>
                	<tr><td>Epha1</td></tr>
                	<tr><td>F12</td></tr>
                	<tr><td>Hgfac</td></tr>
                	<tr><td>Mmp2</td></tr>
                	<tr><td>Mmp9</td></tr>
                	<tr><td>Mylk</td></tr>
                	<tr><td>Pla2r1</td></tr>
                	<tr><td>Prlr</td></tr>
                	<tr><td>Robo1</td></tr>
                	<tr><td>Robo2</td></tr>
                	<tr><td>Tnn</td></tr>
                </table>
                </li>
    		</ol>
    		</li>
    		
    		<li>The InterPro input field supports Boolean operators. (sto199)
 			<p concordion:manual="true">passed. -klf 2014-04-16</p>
			<ol>
				<li>Enter <b>IPR013806 AND NOT IPR000001</b> and submit the query.</li>
				<li>Verify that <b>ONLY the following</b> genes are returned:
				<table>
					<tr><th>Symbol</th></tr>
					<tr><td>Bsph1</td></tr>
					<tr><td>Bsph2</td></tr>
					<tr><td>Fn1</td></tr>
					<tr><td>Igf2r</td></tr>
					<tr><td>Ly75</td></tr>
					<tr><td>Mmp2</td></tr>
					<tr><td>Mmp9</td></tr>
					<tr><td>Mrc1</td></tr>
					<tr><td>Mrc2</td></tr>
					<tr><td>Sel1l</td></tr>
				</table>
				</li>
				<li>Enter <b>calmodulin AND spectrin</b> and submit the query.</li>
				<li>Verify that <b>ONLY the following</b> genes are returned:
				<table>
					<tr><th>Symbol</th></tr>
					<tr><td>Aspm</td></tr>
					<tr><td>Camsap1</td></tr>
					<tr><td>Camsap3</td></tr>
					<tr><td>Vav1</td></tr>
					<tr><td>Vav2</td></tr>
					<tr><td>Vav3</td></tr>
				</table>
				</li>
				<li>Enter <b>Runt OR Opiodes</b> and submit the query.</li>
				<li>Verify that <b>ONLY the following</b> genes are returned:
				<table>
					<tr><th>Symbol</th></tr>
					<tr><td>Pomc</td></tr>
					<tr><td>Runx1</td></tr>
					<tr><td>Runx2</td></tr>
					<tr><td>Runx3</td></tr>
					<tr><td>Trp53</td></tr>
					<tr><td>Trp63</td></tr>
					<tr><td>Trp73</td></tr>
				</table>
				</li>
				<li>step</li>
			</ol>
    		</li>
    		
    		<li>The InterPro input field supports leading and trailing wildcards (asterisk). (consequence of sto199)
 			<p concordion:manual="true">passed. -klf 2014-04-16</p>
			<ol>
				<li>Enter <b>*piod*</b> and submit the query. (Should find the InterPro domain "Opiodes neuropeptide" IPR013532.)</li>
				<li>Verify only <b>Pomc</b> is returned.</li>
			</ol>
    		</li>
    		
    		<li>The InterPro input field supports grouping using parentheses. (consequence of sto199)
 			<p concordion:manual="true">passed. -klf 2014-04-16</p>
			<ol>
				<li>Enter <b>(Runt OR opiodes) AND corticotropin</b> and submit the query.</li>
				<li>Verify only <b>Pomc</b> is returned.</li>
			</ol>
    		</li>
    		
    		<li>The InterPro input field supports phrase matching using double quotation marks. (consequence of sto199)
 			<p concordion:manual="true">passed. -klf 2014-04-16</p>
			<ol>
				<li>Enter <b>"opiomelanocortin/corticotropin"</b> and submit the query.</li>
				<li>Verify only <b>Pomc</b> is returned.</li>
			</ol>
    		</li>
    		
		</ol>

   		<h4 id="mousePheno">Mouse phenotypes &amp; mouse models of human disease</h4>
		<ol>
    		<li>Submitting an MP or OMIM ID returns markers having alleles in OMG or Conditional genotypes associated with the term and its descendants, if any. Markers having alleles only in Complex genotypes associated with the term are excluded.
    		<ol>
                <li>Enter <b concordion:set="#pheno">MP:0009335</b> (decreased splenocyte proliferation) and submit the query. (has no descendants)</li>
                <li>Verify the markers returned include:
	            <table cc:verifySupersetOf="#symbol : getSymbolsByPhenotype(#pheno)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th><th>Notes</th></tr>
                	<tr><td>Aire</td><td></td></tr>
                	<tr><td>Apex2</td><td></td></tr>
                	<tr><td>Il10ra</td><td>Annotated allele is in a Conditional genotype</td></tr>
                	<tr><td>Prkcq</td><td>Annotated alleles in an OMG and a Conditional genotype</td></tr>
                </table>
                </li>
                <li>Verify the following markers are excluded (because their alleles are in Complex genotypes):
	            <table cc:verifyNotRows="#symbol : getSymbolsByPhenotype(#pheno)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th><th>Notes</th></tr>
                	<tr><td>Cd28</td><td></td></tr>
                	<tr><td>Hprt</td><td></td></tr>
                	<tr><td>Impdh1</td><td></td></tr>
                	<tr><td>Tg(Cd4-cre)1Cwi</td><td>Annotated alleles in a Conditional genotype and Type is Recombinase</td></tr>
                </table>
                </li>
    		</ol>
    		</li>

    		<li>Submitting an exact MP or OMIM term returns markers having alleles in OMG or Conditional genotypes associated with the term and its descendants, if any. Markers having alleles only in Complex genotypes associated with the term are excluded.
    		<ol>
                <li>Enter <b concordion:set="#pheno">"decreased splenocyte proliferation"</b> (MP:0009335) USING quotation marks, and submit the query. (has no descendants)</li>
                <li>Verify the markers returned include:
	            <table cc:verifySupersetOf="#symbol : getSymbolsByPhenotype(#pheno)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th><th>Notes</th></tr>
                	<tr><td>Aire</td><td></td></tr>
                	<tr><td>Apex2</td><td></td></tr>
                	<tr><td>Il10ra</td><td>Annotated allele is in a Conditional genotype</td></tr>
                	<tr><td>Prkcq</td><td>Annotated alleles in an OMG and a Conditional genotype</td></tr>
                </table>
                </li>
                <li>Verify the following markers are excluded (because their alleles are in Complex genotypes, or they are a Marker Type that gets excluded):
	            <table cc:verifyNotRows="#symbol : getSymbolsByPhenotype(#pheno)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th><th>Notes</th></tr>
                	<tr><td>Cd28</td><td></td></tr>
                	<tr><td>Hprt</td><td></td></tr>
                	<tr><td>Impdh1</td><td></td></tr>
                	<tr><td>Tg(Cd4-cre)1Cwi</td><td>Annotated allele in a Conditional genotype but Marker Type is a Recombinase.</td></tr>
                </table>
                </li>
    		</ol>
    		</li>

    		<li>Submitting an exact MP or OMIM synonym returns markers having alleles in OMG or Conditional genotypes associated with the term and its descendants, if any.
    		<ol>
                <li>Enter <b concordion:set="#pheno">"reduced splenocyte proliferation"</b> (synonym for MP:0009335 above) USING quotation marks, and submit the query. (has no descendants)</li>
                <li>Verify the markers returned include:
	            <table cc:verifySupersetOf="#symbol : getSymbolsByPhenotype(#pheno)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th><th>Notes</th></tr>
                	<tr><td>Aire</td><td></td></tr>
                	<tr><td>Apex2</td><td></td></tr>
                	<tr><td>Il10ra</td><td>Annotated allele is in a Conditional genotype</td></tr>
                	<tr><td>Prkcq</td><td>Annotated alleles in an OMG and a Conditional genotype</td></tr>
                </table>
                </li>
                <li>Verify the following markers are excluded (because their alleles are in Complex genotypes, or they are a Marker Type that gets excluded):
	            <table cc:verifyNotRows="#symbol : getSymbolsByPhenotype(#pheno)">
        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th><th>Notes</th></tr>
                	<tr><td>Cd28</td><td></td></tr>
                	<tr><td>Hprt</td><td></td></tr>
                	<tr><td>Impdh1</td><td></td></tr>
                	<tr><td>Tg(Cd4-cre)1Cwi</td><td>Annotated allele in a Conditional genotype but Marker Type is a Recombinase.</td></tr>
                </table>
                </li>
    		</ol>
    		</li>

    		<li>Submitting text which appears in more than one MP or OMIM term returns markers having alleles in OMG or Conditional genotypes associated with each term and its descendants, if any.
    		<ol>
                <li>Enter <b concordion:set="#pheno">splenocyte</b> in the Phenotypes/Diseases input text field and submit the search.<br/>(Appears in 11 MP Terms or Synonyms. May also appear in Notes.) </li>
                <li>Verify the markers returned include:
	            <table cc:verifySupersetOf="#symbol : getSymbolsByPhenotype(#pheno)">
								<tr>
									<th concordion:assertEquals="#symbol">Marker Symbol</th>
									<th>Notes</th>
								</tr>
								<tr>
									<td>Aire</td>
									<td></td>
								</tr>
								<tr>
									<td>Apex2</td>
									<td></td>
								</tr>
								<tr>
									<td>Bcl2l11</td>
									<td>Complex geno for some terms, OMG for others, so should
										be included.</td>
								</tr>
								<tr>
									<td>Braf</td>
									<td>MP:0009340, OMGs</td>
								</tr>
								<tr>
									<td>Ccr6</td>
									<td>Has text "splenocyte" in MP annotation notes. Allele
										is "Targeted Reporter" type, but annotation is for OMG
										genotype.</td>
								</tr>
								<tr>
									<td>Cd28</td>
									<td>Has alleles in complex and OMG genotypes.</td>
								</tr>
								<tr>
									<td>Foxp3</td>
									<td>Has text "splenocyte" in MP annotation notes. Allele
										is "Targeted Reporter" type, but annotation is for OMG
										genotype.</td>
								</tr>
								<tr>
									<td>Idd6.1</td>
									<td>Not in annotations, but "splenocyte" exists in (QTL)
										Notes for two QTL alleles.</td>
								</tr>
								<tr>
									<td>Il10ra</td>
									<td>Annotated allele is in a Conditional genotype</td>
								</tr>
								<tr>
									<td>Ncr1</td>
									<td>Has text "splenocyte" in MP annotation notes. Allele
										is "Targeted Reporter" type, but annotation is for OMG
										genotype.</td>
								</tr>
								<tr>
									<td>Prkcq</td>
									<td>Annotated alleles in an OMG and a Conditional genotype</td>
								</tr>
								<tr>
									<td>Srrt</td>
									<td>MP:0009340, Conditional genotype</td>
								</tr>
								<tr>
									<td>Steap3</td>
									<td>MP:0009341, OMG</td>
								</tr>
							</table>
						</li>
                <li>Verify the following markers are excluded (because their alleles are in Complex genotypes, or because of Marker Type):
	            <table cc:verifyNotRows="#symbol : getSymbolsByPhenotype(#pheno)">
								<tr>
									<th concordion:assertEquals="#symbol">Marker Symbol</th>
									<th>Notes</th>
								</tr>
								<tr>
									<td>Hprt</td>
									<td></td>
								</tr>
								<tr>
									<td>Impdh1</td>
									<td></td>
								</tr>
								<tr>
									<td>Casp1</td>
									<td>Only Complex genotypes exist for alleles of this marker.</td>
								</tr>
								<tr>
									<td>H2-D1</td>
									<td>Only Complex genotypes exist for alleles of this marker.</td>
								</tr>
								<tr>
									<td>Igh-J</td>
									<td>Only Complex genotypes exist for alleles of this marker. (Also has OMIM annot for 147060.)</td>
								</tr>
								<tr>
									<td>Tg(Cd4-cre)1Cwi</td>
									<td>Annotated alleles in a Conditional genotype but Marker is a Recombinase, so gets excluded.</td>
								</tr>
							</table>
						</li>
    		</ol>
    		</li>

            <li>Submitting text that is present in a normal MP annotation note -- this should not return the associated gene (5.21).
    		<ol>
                <li>Enter <b concordion:set="#pheno">"notably, supramammillary nuclei"</b> in the Phenotypes/Diseases input text field and submit the search. </li>
               
                <li>Verify the following markers are excluded:
	            <table cc:verifyNotRows="#symbol : getSymbolsByPhenotype(#pheno)">
								<tr>
									<th concordion:assertEquals="#symbol">Marker Symbol</th>
									<th>Notes</th>
								</tr>
								
								<tr>
									<td>Foxb1</td>
									<td>Normal annotation note to MP:0004162 - don't return matches to normal notes</td>
								</tr>
								
							</table>
						</li>
    		</ol>
    		</li>

            <li>Submitting text that is present in a background MP annotation note -- this should return the associated gene.
    		<ol>
                <li>Enter <b concordion:set="#pheno">"the jejunum crypt, but not ileal crypts, are more resistant"</b> in the Phenotypes/Diseases input text field and submit the search. </li>
               
                <li>Verify the following markers are included:
	            <table cc:verifySupersetOf="#symbol : getSymbolsByPhenotype(#pheno)">
								<tr>
									<th concordion:assertEquals="#symbol">Marker Symbol</th>
									<th>Notes</th>
								</tr>
								
								<tr>
									<td>Gpx1</td>
									<td>background sensitivity note</td>
								</tr>
								
							</table>
						</li>
    		</ol>
    		</li>

    		<li>Test that booleans work
    		<ol>
                <li>Query for <b concordion:set="#pheno">MP:0009754 AND MP:0009751</b> should return the following genes:
                  <table cc:verifySupersetOf="#symbol : getSymbolsByPhenotype(#pheno)">
						<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
						<tr><td>Drd4</td></tr>
						<tr><td>hipr</td></tr>
						<tr><td>Ppp1r1b</td></tr>
					</table>
                </li>
                <li>Query for 
                	<b concordion:set="#pheno">"enhanced behavioral response to cocaine" AND "enhanced behavioral response to alcohol"</b> should return the following genes:
                  <table cc:verifySupersetOf="#symbol : getSymbolsByPhenotype(#pheno)">
						<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
						<tr><td>Drd4</td></tr>
						<tr><td>hipr</td></tr>
						<tr><td>Ppp1r1b</td></tr>
					</table>
                </li>
                <li>Query for 
                	<b concordion:set="#pheno">MP:0009754 AND NOT MP:0009751</b> 
                	should return the following genes:
                  <table cc:verifySupersetOf="#symbol : getSymbolsByPhenotype(#pheno)">
						<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
						<tr><td>Camk4</td></tr>
						<tr><td>Chrm4</td></tr>
						<tr><td>Clock</td></tr>
					</table>
					but not these genes:
					<table cc:verifyNotRows="#symbol : getSymbolsByPhenotype(#pheno)">
						<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
						<tr><td>Drd3</td></tr>
						<tr><td>Fyn</td></tr>
						<tr><td>Aldh2</td></tr>
					</table>
                </li>
    		</ol>
    		</li>

    		<li>Test wildcards
    		<ol>
                <li>Query for <b concordion:set="#pheno">Papillorenal</b> should return:
                	<table cc:verifyRowsUnordered="#symbol : getSymbolsByPhenotype(#pheno)">
						<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
						<tr><td>Pax2</td></tr>
					</table>
				</li>
               <li>Query for <b concordion:set="#pheno">Papillore*</b> should return:
                	<table cc:verifyRowsUnordered="#symbol : getSymbolsByPhenotype(#pheno)">
						<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
						<tr><td>Pax2</td></tr>
					</table>
				</li>
				<li>Query for <b concordion:set="#pheno">*apillorenal</b> should return:
                	<table cc:verifyRowsUnordered="#symbol : getSymbolsByPhenotype(#pheno)">
						<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
						<tr><td>Pax2</td></tr>
					</table>
				</li>
    		</ol>
    		</li>

    		<li>Transgenic Reporter markers are excluded from the results set.
 			<p concordion:manual="true">passed. -klf 3/31/2014</p>
    		<ol>
                <li>Query for Medulloblastoma; MDB <b concordion:set="#pheno">155255</b>.</li>
                <li>Verify that <b>Ptch1</b> is one of the markers returned. (In some OMGs and also Complex genotype with Transgenic Reporter below.)</li>
                <li>Verify that <b>Tg(Atoh1-GFP)1Jejo</b> is not returned. (Transgenic Reporter portion of the genotype MGI:3836962)</li>
                <li>Query for Teratoma, Ovarian <b concordion:set="#pheno">166950</b>.</li>
                <li>Verify that <b>Foxo3</b> is one of the markers returned.</li>
                <li>Verify that <b>Tg(HBA1-Gfp)1Ew</b> is not returned. (Transgenic Reporter portion of the genotype MGI:4949205)</li>
    		</ol>
    		</li>

    		<li>Markers associated with Recombinase alleles are excluded from the results set.
 			<p concordion:manual="true">passed. -klf 3/31/2016</p>
    		<ol>
                <li>Included in tests above.</li>
    		</ol>
    		</li>
    		
		</ol>

		<h4 id="sorting">Sorting</h4>
    	<ol>	
    		<li>Default sort order when Nomenclature field is submitted reflects the table in the Logic section as much as possible.
 			<p concordion:manual="true">passed - kstone 2014/04/23</p>
 			<ol>
 				<li>Submit a query with the data specified in the Nomen field below.</li>
 				<li>Verify that the order of results returned reflects the table in the Logic section as much as possible.</li>
 				<li>
 				<table>
<!--
select label, count(*)
from MRK_Label
group by label
order by count(*) desc  
-->
						<tr>
							<th>Nomen</th>
							<th>Notes to tester</th>
						</tr>
						<tr>
							<td>fp</td>
							<td>matches current name, old symbol, synonym, human
								synonym, or allele symbol for several genes.<br /> Brings
								back Srrm4, Ptgfr, Sdha like Public. Excludes many Public
								"contains" matches. Includes 5 markers not returned by Public.
							</td>
						</tr>
						<tr>
							<td>cat</td>
							<td>matches current symbol, current name, old symbol,
								synonym, allele symbol, or human synonym for many genes.<br />
								Brings back Cat, Mip, Cac, Trpv6, Glyat, Git1, Git2, Cat2,
								Cat3 like Public, but different order!<br /> Excludes many
								Public "contains" matches.
							</td>
						</tr>
						<tr>
							<td>C57BL/6J</td>
							<td>Public returns 1705 markers, mostly due to Allele
								Name, 25 for Allele Symbol.<br /> Valinor returns 1674
								markers, 3 for Current Name, all the rest for Allele Symbol.
							</td>
						</tr>
						<tr>
							<td>c4a</td>
							<td>Returns 3 matches on Valinor, 26 matches on Public due
								to "contains".<br /> Valinor matches due to Current Symbol,
								Current Name, or Human Symbol.
							</td>
						</tr>
						<tr>
							<td>ob</td>
							<td>Public returns 103 markers; Valinor returns 10.<br />
								Differences: Meiob and H2-I are returned 2nd on Valinor due to
								Current Name.<br /> The top 3 returns on Public are in the top
								6 on Valinor.
							</td>
						</tr>
						<tr>
							<td>scid</td>
							<td>Public returns 1 more gene than Valinor due to "contains" searching.</td>
						</tr>
					</table>
				</li>
 			</ol>
    		</li>
    			
    		<li>Default sort order when Feature Type rather than the Nomenclature field is submitted is MGI's "smart alpha" by Marker Symbol.
 			<p concordion:manual="true">passed. -klf 4/17/2014</p>
 			<ol>
 				<li>Select only <b>unclassified other genome feature</b> and submit the query.</li>
 				<li>Verify that markers beginning with <b>Defa-...</b> are in "smart alpha" order.</li>
 			</ol>
    		</li>

    		<li>Default sort order when Map Position rather than the Nomenclature field is submitted is MGI's "smart alpha" by Marker Symbol.
 			<p concordion:manual="true">passed. -klf 4/17/2014</p>
 			<ol>
 				<li>Select chromosome <b>Y</b> and submit the query.</li>
 				<li>Verify that markers beginning with <b>DYBish...</b> are in "smart alpha" order.</li>
 			</ol>
    		</li>

    		<li>Default sort order when the Nomenclature field and a field other than the Nomenclature field is submitted follows the table in the Logic section, as much as possible.
 			<p concordion:manual="true">passed. -klf 4/17/2014</p>
 			<ol>
 				<li>Enter <b>cytochrome</b> in the Nomenclature input field and submit the query.</li>
 				<li>Note the order of the <b>cytochrome P450</b> families of markers in the output.</li>
 				<li>Enter <b>cytochrome</b> in the Nomenclature input field and <b>monooxygenase</b> in the GO input field and submit the query.</li>
 				<li>Of the markers that are returned, verify that the <b>cytochrome P450</b> families of markers are in the same order. (<em>Note: not all will be returned because the GO term filters some markers out.</em>)</li>
 			</ol>
    		</li>
    			
    		<li>On-demand sort by clicking the "Symbol" column header in Summary output resorts all returned data in "smart alpha" order by Symbol.
 			<p concordion:manual="true">passed. -klf 4/17/2014</p>
 			<ol>
 				<li>Select chromosome <b>Y</b> and submit the query.</li>
 				<li>Click the <b>Symbol</b> column header.</li>
 				<li>Verify that markers are in "smart alpha" order.</li>
 			</ol>
    		</li>

    		<li>On-demand sort by clicking the "Genetic Location" column header in Summary output resorts all returned data in by Chromosome, then Start Coordinate, then Symbol.
 			<p concordion:manual="true">passed. -klf 4/17/2014</p>
 			<ol>
 				<li>Enter <b>cytochrome</b> in the Nomenclature input field and submit the query.</li>
 				<li>Click the <b>Genetic Location</b> column header.</li>
 				<li>Verify that the <b>cytochrome P450</b> families of markers are in order by chromosome, then start coordinate, then symbol by "smart alpha".</li>
 			</ol>
    		</li>
    		
    		<li>On-demand sort by clicking the "Feature Type" column header in Summary output resorts all returned data in "smart alpha" order by Feature Type(sto203).
 			<p concordion:manual="true">passed. -jeffc 4/28/2014</p>
 			<ol>
 				<li>Select Gene/Marker Symbol/Name <b>Pax*</b> and submit the query.</li>
 				<li>Click the <b>Feature Type</b> column header.</li>
 				<li>Verify that Feature Type are in "smart alpha" order.</li>
 			</ol>
    		</li>
   		</ol>

		<h4 id="youSearchedFor">You Searched For...</h4>
		<ol>
		<li>Summary displays "You searched for..." reiterating the user's input for each submitted parameter. Parameters are listed in the same order as on the query form.  The first parameter submitted is listed with the label specified. The word "AND" precedes all subsequent parameter labels to indicate the query logic. Parameters which may have multiple values display all selected or input values. (sto168)
 			<p concordion:manual="true">passed. klf 4/11/2014</p>
 			<ol>
 				<li>Submit a Marker Query with the following parameters entered or selected:
 				<ul>
 					<li>Marker Symbol/Name: <b>Pa*</b></li>
 					<li>Feature type: <b>gene</b></li>
 					<li>Map position: Chromosomes <b>1,2,3,4,5,6,7,8,9</b></li>
 					<li>Genome Coordinates: <b>1-999999999999</b> bp</li>
 					<li>Gene Ontology (GO) classifications: <b>cell</b>; select <b>Molecular Function</b> and <b>Cellular Component</b></li>
 					<li>InterPro Protein Domains: <b>domain</b></li>
 					<li>Phenotype/Human Disease: <b>abnormal OR phenotype</b></li>
 				</ul>
 				</li>
 				<li>Verify the results display includes the "You searched for" section with the submitted parameters listed in the order shown above.</li>
 				<li>Verify the "You searched for" section shows each submitted parameter value with its label and "help text" where needed:
 				<ul>
 					<li>Marker Symbol/Name: including text <b>Pa*</b> <span style="font-size:0.75em; color:#999999;">searching current symbols/names and synonyms</span></li>
 					<li>AND Feature type: any of <b>[gene, protein coding gene, non-coding RNA gene, rRNA gene, ...]</b> <em>(NOTE: We have to show all "child terms" of the selected term too. Need feedback from the Demo to see if we should substitute generic text like "your selected ")</em></li>
 					<li>AND on Chromosome: any of <b>[1, 2, 3, 4, 5, 6, 7, 8, 9]</b></li>
 					<li>AND Genome Coordinates: <b>1-999999999999</b> bp</li>
 					<li>AND Gene Ontology term(s): contains <b>cell</b> searching <b>Molecular Function</b> and <b>Cellular Component</b></li>
 					<li>AND InterPro Protein Domains: contains <b>domain</b></li>
 					<li>AND Phenotype/Human Disease: including text <b>abnormal OR phenotype</b> <span style="font-size:0.75em; color:#999999;">searching MP terms, synonyms, IDs, and notes, disease terms, synonyms and IDs</span></li>
 				</ul>
 				</li>
 				<li>Submit a Marker Query with the following parameters entered or selected:
 				<ul>
 					<li>Marker Symbol/Name: <b>Pa*</b></li>
 					<li>Feature type: <b>other feature type</b></li>
 					<li>Map position: Chromosomes <b>10, 11, 12, 13, 14, 15, 16, 17, 18, 19, X, Y, XY, Unknown, Mitochondrial</b></li>
 					<li>Genome Coordinates: <b>1-999</b> Mbp</li>
 					<li>Gene Ontology (GO) classifications: <b>cell</b> (with no GO sub-DAGs selected)</li>
 					<li>InterPro Protein Domains: <b>domain</b></li>
 					<li>Phenotype/Human Disease: <b>abnormal OR phenotype</b></li>
 				</ul>
 				</li>
 				<li>Verify the "You searched for" section shows each submitted parameter value with its label and "help text" where needed:
 				<ul>
 					<li>Marker Symbol/Name: including text <b>Pa*</b> <span style="font-size:0.75em; color:#999999;">searching current symbols/names and synonyms</span></li>
 					<li>AND Feature type: any of <b>[other feature type, QTL, transgene, complex/cluster/region, ...]</b> <em>(NOTE: We have to show all "child terms" of the selected term too. Need feedback from the Demo to see if we should substitute generic text like "your selected ")</em></li>
 					<li>AND on Chromosome: any of <b>[10, 11, 12, 13, 14, 15, 16, 17, 18, 19, X, Y, XY, UN, MT]</b></li>
 					<li>AND Genome Coordinates: <b>1-999</b> Mbp</li>
 					<li>AND Gene Ontology term(s): contains <b>cell</b></li>
 					<li>AND InterPro Protein Domains: contains <b>domain</b></li>
 					<li>AND Phenotype/Human Disease: including text <b>abnormal OR phenotype</b> <span style="font-size:0.75em; color:#999999;">searching MP terms, synonyms, IDs, and notes, disease terms, synonyms and IDs</span></li>
 				</ul>
 				</li>
 				<li>Submit a Marker Query with the following parameters entered or selected:
 				<ul>
 					<li>Marker Symbol/Name: <em>default setting = blank</em></li>
 					<li>Feature type: <em>default setting</em></li>
 					<li>Map position: Chromosomes <b>1</b></li>
 					<li>cM Position: <b>10.0-40.0</b></li>
 					<li>Gene Ontology (GO) classifications: <b>cell</b> (with no GO sub-DAGs selected)</li>
 					<li>InterPro Protein Domains: <em>default setting = blank</em></li>
 					<li>Phenotype/Human Disease: <em>default setting = blank</em></li>
 				</ul>
 				</li>
 				<li>Verify the "You searched for" section shows each submitted parameter value with its label and "help text" where needed:
 				<ul>
	 				<li>on Chromosome: any of <b>[1]</b></li>
 					<li>AND cM Position: <b>10.0-40.0</b></li>
 					<li>AND Genome Coordinates: 1-999 Mbp </li>
 					<li>AND Gene Ontology term(s): contains <b>cell</b> </li>
 				</ul>
 				</li>
 			</ol>
		</li>
		
		<li>"You searched for..." does not reiterate default values submitted for any parameter. (sto168)
 			<p concordion:manual="true">passed. klf 4/11/2014</p>
 			<ol>
 				<li>Navigate to the Marker Query Form.</li>
 				<li>Submit a Marker Query with the following parameters entered or selected:
 				<ul>
 					<li>Marker Symbol/Name: <em>default setting = blank</em></li>
 					<li>Feature type: <em>default setting</em></li>
 					<li>Map position: Chromosomes <em>default setting</em></li>
 					<li>Marker Range: between <b>D19Mit32</b> and <b>Tbx10</b></li>
 					<li>Gene Ontology (GO) classifications: <b>cell</b> (with no GO sub-DAGs selected)</li>
 					<li>InterPro Protein Domains: <em>default setting = blank</em></li>
 					<li>Phenotype/Human Disease: <em>default setting = blank</em></li>
 				</ul>
 				</li>
 				<li>Verify the "You searched for" section shows ONLY each submitted parameter value with its label and "help text" where needed and does NOT show default values for Feature type, Chromosome or GO sub-DAGs: <span concordion:manual="true">passed. -klf 4/11/2014</span>
 				<ul>
 					<li>Marker Range: between <b>D19Mit32</b> and <b>Tbx10</b> </li>
 					<li>AND Gene Ontology term(s): contains <b>cell</b> </li>
 				</ul>
 				</li>
 				<li>Use the <b>Reset</b> button to clear values from the Marker Query Form. </li>
 				<li>Submit a Marker Query with the following parameters entered or selected:
 				<ul>
 					<li>Marker Symbol/Name: <em>default setting = blank</em></li>
 					<li>Feature type: <em>default setting</em></li>
 					<li>Map position: <em>default setting = blank</em></li>
 					<li>Gene Ontology (GO) classifications: <em>default setting = blank</em></li>
 					<li>InterPro Protein Domains: <em>default setting = blank</em></li>
 					<li>Phenotype/Human Disease: <b>abnormal AND phenotype</b></li>
 				</ul>
 				</li>
 				<li>Verify the "You searched for" section shows ONLY each submitted parameter value with its label and "help text" where needed and does NOT show default values for Feature type, Chromosome or GO sub-DAGs: <span concordion:manual="true">passed. -klf 4/11/2014</span>
 				<ul>
	 				<li>Phenotypes/Diseases: including text <b>abnormal AND phenotype</b> <span style="font-size:0.75em; color:#999999;">searching MP terms, synonyms, IDs, and notes, disease terms, synonyms and IDs</span></li>
 				</ul>
 				</li>
 			</ol>
		</li>
		
		</ol>   		

		<h4 id="downloadableFiles">Downloadable Files</h4>
		<ol>
			<li>A file containing all data in the current results set can be downloaded in Text (tab-delimited) or Excel format.
 			<p concordion:manual="true">passed. -klf 4/11/2014</p>		
 			<ol>
 				<li>Enter <b>Pax6</b> in the Gene/Marker field and submit the query.</li>
 				<li>Click the <b>Text File</b> button.</li>
 				<li>Verify that 18 rows are returned.</li>
 				<li>Verify that the filename contains the current date and current time.</li>
 				<li>Return to the query results and click the <b>Excel File</b> button.</li>
 				<li>Verify that 18 rows are returned.</li>
 				<li>Verify that the filename contains the current date and current time.</li>
 				<li>Reset the query form. Enter <b>GO:0071314</b> (cellular response to cocaine) in the Gene Ontology classifications field and submit the query.</li>
 				<li>Click the <b>Text File</b> button.</li>
 				<li>Verify that 2 rows are returned.</li>
 				<li>Verify that the filename contains the current date and current time.</li>
 				<li>Return to the query results and click the <b>Excel File</b> button.</li>
 				<li>Verify that 2 rows are returned.</li>
 				<li>Verify that the filename contains the current date and current time.</li>
 				 
 			</ol>
			</li>

			<li>Downloads have the correct data columns: (sto164)
 			<p concordion:manual="true">passed. -klf 4/11/2014</p>		
			<table>
				<tr><td>Chr <em>(NOTE: this is genomic chromosome.  No differences from genetic chr exist as of 4/10/14 so genetic chr was dropped as redundant.)</em></td></tr> 
				<tr><td>start </td></tr>
				<tr><td>end </td></tr>
				<tr><td>cM <em>(Note: the web iterface translates some special values to text: "QTL", "syntenic", maybe others. This was dropped from the file download.)</em></td></tr>
				<tr><td>strand GRCm39</td></tr> 
				<tr><td>MGI ID </td></tr>
				<tr><td>Feature Type</td></tr> 
				<tr><td>Symbol </td></tr>
				<tr><td>Name </td></tr>
				<tr><td>Matching Text (optional - Matching Text appears when the user submits a nomen query)</td></tr>
			</table>
 			<ol>
 				<li>Using the Text and Excel files returned for the <b>Pax6</b> query above, verify that these columns are returned, including <b>Why Matched?</b>.</li>
 				<li>Verify that the data in the downloaded file matches the data displayed in the query results page. </li>
 				<li>Using the Text and Excel files returned for the <b>GO:0071314</b> query above, verify that these columns are returned, including <b>Why Matched?</b>.</li>
 				<li>Verify that the data in the downloaded file matches the data displayed in the query results page. </li>
 			</ol>
			</li>
		
			<li>The cM column value is a number between 1 and 110 or is blank. <em>(From Schema Browser: Offset value of -1.0 designates syntenic marker. Offset value of -999.9 designates withdrawn or reserved marker.)</em> (sto164)
 			<p concordion:manual="true">passed. -klf 4/11/2014</p>
 			<ol>
 				<li>Query for Map Position Marker Range between <b>Hrh3</b> and <b>Vacq1</b>.</li>
 				<li>Verify that the results set includes markers with Genetic Location including "syntenic", "QTL", and values greater than 100.</li>
 				<li>Click the Text File button.</li>
 				<li>Verify that NO cM value for <b>1700039O17Rik</b> is displayed. <em>(Marker is syntenic; the cM value in the database is -1.0.)</em></li>
 				<li>Verify that the cM value for <b>4921531C22Rik</b> is 102.49. <em>(Marker has a cM value.)</em></li>
 				<li>Verify that the cM value for <b>Alcp5</b> is 103.43. <em>(Marker is a QTL with a cM value.)</em></li>
 				<li>Click the Excel File button.</li>
 				<li>Verify that </li>
 				<li>Step</li>
 			</ol>
 			</li>
		
			<li>The "Why Matched?" column value does not contain HTML "b" tags surrounding the matching text portion of the string. (sto164)
 			<p concordion:manual="true">passed. -klf 4/11/2014</p>
 			<ol>
 				<li>Using the Text and Excel files returned for the <b>Pax6</b> query above, verify that the <b>Why Matched?</b> column does not contain "&lt;b&gt;" or "&lt;/b&gt;" tags.</li>
 			</ol>
 			</li>
 			
			<li>A text search in the Pheno/Disease field ignores TEXT-QTL mapping experiment notes unless it for a marker of type QTL (sto215)
 			<p concordion:manual="true">passed. -Jeffc 4/24/2014</p>
 			<ol>
 				<li>Using the Marker Query form</li>
 				<li>Enter <b>Breast Cancer</b> in the Phenotype/Human Disease field</li>
 				<li>Verify the results do not include any Riken genes, since these genes are associated with a TEXT-QTL Mapping Experiment and the only text match is in the Mapping Experiment Note (they don't have any matches in MP or OMIM annotated data).</li>
 			</ol>
 			</li>		
		</ol>

	</div>

	<div id="auto">
		<h3>Automated tests</h3>
		<ol>
    		<li>Nomen Query
    		<ol>
               	<li>Testing that AND, OR, and NOT aren't treated as Booleans
				<ol>
					<li>open the Marker query form</li>
					<li>Type <b concordion:set="#query">pax6 and shh</b> in the nomenclature field and press the search button</li>
					<li>verify that <b concordion:assertEquals="getMarkerCountByNomen(#query)">0</b> matching genes are returned</li>
					<li>Type <b concordion:set="#query">pax6 or shh</b> in the nomenclature field and press the search button</li>
					<li>verify that <b concordion:assertEquals="getMarkerCountByNomen(#query)">0</b> matching genes are returned</li>
					<li>Type <b concordion:set="#query">paired AND box</b> in the nomenclature field and press the search button</li>
					<li>verify that <b concordion:assertEquals="getMarkerCountByNomen(#query)">0</b> matching genes are returned</li>
					<li>Type <b concordion:set="#query">paired OR box</b> in the nomenclature field and press the search button</li>
					<li>verify that <b concordion:assertEquals="getMarkerCountByNomen(#query)">0</b> matching genes are returned</li>
					<li>Type <b concordion:set="#query">paired NOT box</b> in the nomenclature field and press the search button</li>
					<li>verify that <b concordion:assertEquals="getMarkerCountByNomen(#query)">0</b> matching genes are returned</li>
				</ol>
				</li>
				<li>Testing that punctuation and extraneous wildcards are removed (ignored)
				<ol>
					<li>open the Marker query form</li>
					<li>Type <b concordion:set="#query">dopachrome@#$%^&amp;*()_-+={}[]\|`~;':",./&lt;&gt;?</b> in the nomenclature field and press the search button</li>
					<li>verify that only <b concordion:assertEquals="getMarkerCountByNomen(#query)">2</b> markers returned</li>
					<li>Type <b concordion:set="#query">*dopachrome ***tautome*</b> in the nomenclature field and press the search button</li>
					<li>verify that only <b concordion:assertEquals="getMarkerCountByNomen(#query)">2</b> markers is returned</li>
					<li>Type <b concordion:set="#query">*****</b> in the nomenclature field and press the search button</li>
					<li>verify that <b cc:assertEquals="getMarkerCountByNomen(#query)">0</b> matching markers are returned</li>
				</ol>
				</li>
    		</ol>
    		</li>
    		<li>Feature Type Query
    		<ol>
               	<li>Querying all feature types should bring back all genes
				<ol>
					<li>open the Marker query form</li>
					<li>Select <b>all feature types</b> and search</li>
					<li>Summary should return all <b cc:assertEquals="getMarkerCountAllMcv()"><b cc:dynamicData="allMarkers"></b></b></li>
				</ol>
				</li>
				<li>Markers returned for the selected feature type are of that type
	    		<ol>
	                <li>Select <b>polymorphic pseudogene</b> (count is 4 as of 3/25/2014) and submit the query with no other parameters.
	                 <span class="hide" concordion:set="#mcvKey">7288449</span></li>
	                <li>Verify <b cc:assertEquals="getMarkerCountByMcv(#mcvKey)"><b cc:dynamicData="markerCountMcv1"></b></b> markers are returned.</li>
	                <li>Verify the markers returned are all of the selected type.</li>
	                <li>Reopen the query and add selections for <b>RNase MRP RNA gene</b> (count=1) and <b>telomerase RNA gene</b> (count=1) and submit the query.
	                	<span class="hide" concordion:set="#mcvKey">7288449,6238182,6238183</span></li>
	                <li>Verify <b cc:assertEquals="getMarkerCountByMcv(#mcvKey)"><b cc:dynamicData="markerCountMcv2"></b></b> markers are returned.</li>
	                <li>Verify the 2 additional markers are of the additional types just selected.</li>
	    		</ol>
	    		</li>
    		</ol>
    		</li>
    		<li>Map Position
    		<ol>
	   			<li>Submitting a single chromosome selection returns markers for the chromosome.  Submitting multiple chromosome selections returns markers for all selected chromosomes. 
	    		<ol>
	                <li>Select chromosome <b concordion:set="#chr">19</b> and submit the query with no other parameters.</li>
	                <li>Verify <b cc:assertEquals="getMarkerCountByChromosome(#chr)"><b cc:dynamicData="markerCountChr19"></b></b> markers are returned.</li>
	                <li>Reopen the query select only chromosome <b concordion:set="#chr">Y</b> and submit the query.</li>
	                <li>Verify <b cc:assertEquals="getMarkerCountByChromosome(#chr)"><b cc:dynamicData="markerCountChrY"></b></b> markers are returned.</li>
	                <li>Reopen the query select chromosomes <b concordion:set="#chr">19, Y</b> and submit the query.</li>
	                <li>Verify <b cc:assertEquals="getMarkerCountByChromosome(#chr)"><b cc:dynamicData="markerCountChr19andY"></b></b> markers are returned.</li>
	    		</ol>
	   			</li>

	    		<li>Submitting a chromosome + cM position range returns markers in that range.
	    		<ol>
	                <li>Select chromosome <b concordion:set="#chr">1</b> and cM range <b concordion:set="#cm">10-20</b> and submit the query with no other parameters.</li>
	                <li>Verify the list includes the following markers:
	                <table cc:verifySupersetOf="#symbol : getSymbolsByChrCm(#chr,#cm)">
	        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
	                	<tr><td>Aff3</td></tr>
	                	<tr><td>Cocia4</td></tr>
	                	<tr><td>Creg2</td></tr>
	                	<tr><td>D1Mit20</td></tr>
	                	<tr><td>Eif5b</td></tr>
	                	<tr><td>Gm4850</td></tr>
	                	<tr><td>Prss40</td></tr>
	                	<tr><td>Snord89</td></tr>
	                	<tr><td>Zfp451</td></tr>
	                </table>
	                </li>
	    		</ol>
	    		</li>
	
	    		<li>Submitting a chromosome + genome coordinates range returns markers in that range.
	    		<ol>
	                <li>Select chromosome <b concordion:set="#chr">1</b> and genome coordinates range <b concordion:set="#coord">24332803-40772038</b> with units <b>bp</b> and submit the query with no other parameters.</li>
	                <li>Verify the list includes the following markers:
	                <table cc:verifySupersetOf="#symbol : getSymbolsByChrCoord(#chr,#coord)">
	        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
	                	<tr><td>Aff3</td></tr>
	                	<tr><td>Creg2</td></tr>
	                	<tr><td>D1Mit20</td></tr>
	                	<tr><td>Eif5b</td></tr>
	                	<tr><td>Gm4850</td></tr>
	                	<tr><td>Prss40</td></tr>
	                	<tr><td>Snord89</td></tr>
	                	<tr><td>Zfp451</td></tr>
	                </table>
	                </li>
	                <li>Verify the following markers are not returned (no coordinates):
	                <table cc:verifyNotRows="#symbol : getSymbolsByChrCoord(#chr,#coord)">
	        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
	                	<tr><td>Cocia4</td></tr>
	                	<tr><td>D1Mit412</td></tr>
	                	<tr><td>Dbsty1</td></tr>
	                </table>
	                </li>
	                <li>Reopen the query and select chromosome <b concordion:set="#chr">1</b> 
	                	and genome coordinates range <b concordion:set="#coord">24.332-40.772</b> with units <b>Mbp</b> and submit the query with no other parameters.</li>
	                <li>Verify the following markers are returned:
	                <table cc:verifySupersetOf="#symbol : getSymbolsByChrCoordMbp(#chr,#coord)">
	        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
	                	<tr><td>Aff3</td></tr>
	                	<tr><td>Amer3</td></tr>
	                	<tr><td>Bag2</td></tr>
	                </table>
	                </li>
	    		</ol>
	    		</li>
	    		<li>Submitting a marker range returns markers which fall between the start coordinate of the one marker and the end coordinate of the other marker.
	    		<ol>
	                <li>Input <b concordion:set="#startMarker">D19Mit32</b> and <b concordion:set="#endMarker">Tbx10</b>.</li>
	                <li>Verify the following markers are returned (these are sorted by genome coordinates):
	                <table cc:verifySupersetOf="#symbol : getSymbolsByRange(#startMarker,#endMarker)">
	        			<tr><th concordion:assertEquals="#symbol">Marker Symbol</th></tr>
	                	<tr><td>Ighmbp2</td></tr>
						<tr><td>D19Mit32</td></tr>
						<tr><td>Lfp3</td></tr>
						<tr><td>Tgct2</td></tr>
						<tr><td>Mrpl21</td></tr>
						<tr><td>Cpt1a</td></tr>
						<tr><td>4833417J20Rik</td></tr>
						<tr><td>Mtl5</td></tr>
						<tr><td>Gal</td></tr>
						<tr><td>Ppp6r3</td></tr>
						<tr><td>Gm23940</td></tr>
						<tr><td>Gm19205</td></tr>
						<tr><td>n-TAagc9</td></tr>
						<tr><td>Lrp5</td></tr>
						<tr><td>D19Mit68</td></tr>
						<tr><td>Drinksac2</td></tr>
						<tr><td>1810055G02Rik</td></tr>
						<tr><td>Gm2141</td></tr>
						<tr><td>Gm16066</td></tr>
						<tr><td>Gm19209</td></tr>
						<tr><td>Chka</td></tr>
						<tr><td>Tcirg1</td></tr>
						<tr><td>Ndufs8</td></tr>
						<tr><td>Aldh3b1</td></tr>
						<tr><td>Unc93b1</td></tr>
						<tr><td>Aldh3b2</td></tr>
						<tr><td>Aldh3b3</td></tr>
						<tr><td>Acy3</td></tr>
						<tr><td>Tbx10</td></tr>
	                </table>
	                </li>
	    		</ol>
	    		</li>
	    		
    		</ol>
    		</li>
		</ol>
 	</div>
</div>

</body>
</html>
